terraform/modules/hetzner-kubernetes/machines.resources.tf

```diff
@@ -0,0 +1,54 @@
+resource "hcloud_server" "machines" {
+  for_each = { for m in var.machines: "${m.group_name}-${m.machine_id}" => m }
+
+  name        = "${var.cluster_name}-${each.key}"
+  location    = var.default_location
+  image       = var.default_image
+  server_type = lookup(each.value, "machine_type", var.default_server_type)
+  ssh_keys    = var.ssh_keys
+
+  # NOTE: string is the only accepted type
+  # DOCS: for possible characters, see https://docs.hetzner.cloud/#labels
+  labels = merge(
+    {
+      cluster = var.cluster_name
+      group_name = each.value.group_name
+      machine_id = each.value.machine_id
+    },
+    contains( each.value.component_classes, "controlplane" ) ? { etcd_member_name = "etcd-${ each.value.machine_id }" } : {},
+    { for class in each.value.component_classes : "component-class.${local.LABEL_PREFIX}/${class}" => true }
+  )
+}
+
+
+resource "hcloud_server_network" "snw" {
```

Comment:
I think this should preferably not live in the kubernetes module, but in a separate `private_network` module and then be an input to the kubernetes module.

Say I have another module defining `cassandra` cluster or `elasticsearch` cluster; I want these to be in the same private network as the kubernetes cluster.  (Because I want kubernetes nodes to be able to talk to these services)

Another option would be to let the `kubernetes` module provide the private network as an output, but this is harder for refactoring and a bit awkward (Why would the `cassandra` module have to depend on the `kubernetes` module; for example)

See Also  the section about [Dependency Inversion](https://www.terraform.io/docs/language/modules/develop/composition.html#dependency-inversion) in the modules best practises guide.

Reply:
The scenario you are describing was exactly @jmatsushita and my reasoning behind introducing `machine_groups`. This allows to create only K8s groups (controlpales & nodes) but also other groups in addition, by deriving the Ansible groups from whatever items are defined in `component_classes`. Meaning a item of `component_classes` equals to an Ansible inventory group. This way, all machines would share the same private network.

Reply:
Hmm I really feel like we're trying to re-invent the wheel here a bit too much still.  What doe we gain from using `machine_groups` in `tfvars` over defining a `resource` (or a `module`)  in terraform code? We're coming up with our own DSL for defining machine groups; which terraform itself already has...    Defining how we provision `cassandra` and elasticsearch groups  whilst calling the `kubernetes` module also sounds somewhat confusing to me.    I am still not really fond of this idea.

Reply:
Why I am not fond of it:

The terraform code alone will not describe the topology of a Wire installation anymore. There will just be one node in the graph called "machine" and it's up to the user to define a topology in their `tfvars` that makes sense.

If we instead have explicit `cassandra`, `kubernetes` and `elasticsearch` `hcloud_server` resources (or modules if there is a group of resources you want to abstract) then it's clear from the terraform code what a wire deployment topology looks like. 

This means we have to explicitly document how people should configure this topology in `tfvars` instead of having it evident in the code itself.